core/src/main/scala/kafka/coordinator/transaction/TransactionCoordinator.scala

```diff
@@ -383,11 +384,43 @@ class TransactionCoordinator(brokerId: Int,
 
   def partitionFor(transactionalId: String): Int = txnManager.partitionFor(transactionalId)
 
+  private def expireTransactions(): Unit = {
+
+    txnManager.transactionsToExpire().foreach{ idAndMetadata =>
+      idAndMetadata.metadata synchronized {
+        if (!txnManager.isCoordinatorLoadingInProgress(idAndMetadata.transactionalId)) {
+          idAndMetadata.metadata.producerEpoch = (idAndMetadata.metadata.producerEpoch + 1).toShort
+          idAndMetadata.metadata.prepareTransitionTo(Ongoing)
+          txnManager.appendTransactionToLog(idAndMetadata.transactionalId, idAndMetadata.metadata, (errors: Errors) => {
```

Comment:
Why can't we initiate an ABORT directly with the bumped epoch? We have to write the `PREPARE` anyway. We might as well do it with the the higher epoch.